Initial work on new operational mode page (#2188)
hashicorp/terraform-website: content/enterprise/operational-modes/index.mdx

- **Vault:** - Encryption keys that encrypt and decrypt objects within object storage.

- **Redis:** - Application coordination and data caching.

- **Configuration:** - Configuration settings. This includes the hostname, object storage credentials, database credentials, concurrency settings, etc.

## Operational Modes

The following table summarizes where each Terraform Enterprise operational mode
stores each type of application data.

| Operational Mode                        | PostgreSQL Database                                                  | Object Storage                                                      | Vault                                                                                           | Redis                                  | Configuration                  |
| --------------------------------------- | -------------------------------------------------------------------- | ------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------- | -------------------------------------- |------------------------------- |

> | [External Services](#external-services) | External, user-managed PostgreSQL database.                          | External, user-managed object storage location.                     | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault). | Docker volume on the instance.         | Docker volume on the instance. |
> | [Active/Active](#active-active)         | External, user-managed PostgreSQL database.                          | External, user-managed object storage location.                     | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault). | External, user-managed Redis instance. | Docker volume on the instance. |
> | [Mounted Disk](#mounted-disk)           | Directory on the instance backed by user-managed persistent storage. | Directory on the instance backed by user-managed persisent storage. | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault). | Docker volume on the instance.         | Docker volume on the instance. |

Reviewer: ```suggestion
| [External Services](#external-services) | External, user-managed PostgreSQL database                          | External, user-managed object storage location                     | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault) | Docker volume on the instance         | Docker volume on the instance |
| [Active/Active](#active-active)         | External, user-managed PostgreSQL database                          | External, user-managed object storage location                     | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault) | External, user-managed Redis instance | Docker volume on the instance |
| [Mounted Disk](#mounted-disk)           | Directory on the instance backed by user-managed persistent storage | Directory on the instance backed by user-managed persisent storage | PostgreSQL Database unless using [External Vault](/enterprise/requirements/data-storage/vault) | Docker volume on the instance         | Docker volume on the instance |
```

## External Services

The External Services operational mode stores the PostgreSQL Database and
Object Storage data on an external PostgreSQL server and S3-compatible object
storage. This allows you to backup, restore, and scale the PostgreSQL Database
or Object Storage independently of the Terraform Enterprise instance.

You are responsible for providing and maintaining the external PostgreSQL
server and S3-compatible object storage.

> Configuration data is stored on a Docker volume on the instance. You can
> retrieve this information using `replicatedctl app-config export`.

Reviewer: ```suggestion
External Services stores configuration data on a Docker volume on the instance. Use `replicatedctl app-config export` to retrieve the configuration data.
```

### When to Use External Services

> External Services is recommended when:

Reviewer: ```suggestion
We recommend using External Services in the following use cases:
```

> - Running Terraform Enterprise in environment that has an external PostgreSQL server and S3-compatible object storage.

Reviewer: ```suggestion
- Your run environment has an external PostgreSQL server and S3-compatible object storage.
```

- You have experience managing a PostgreSQL server such as Amazon RDS for PostgreSQL, Google Cloud SQL for PostgreSQL, Azure Database for PostgreSQL, etc.
- You have experience managing an S3-compatible object storage location such as AWS S3, Azure Blob Storage, Google Cloud Storage, MinIO, etc.

> - Want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.

Reviewer: ```suggestion
- You want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.
```

## Active/Active

The Active/Active operational mode stores the PostgreSQL Database and Object
Storage data on an external PostgreSQL server and S3-compatible object storage.
This allows you to backup, restore, and scale the PostgreSQL Database or Object

> Storage independently of the Terraform Enterprise instance. Additionally, Redis
> data is stored in an external Redis instance, allowing for multiple instances
> of Terraform Enterprise to be used in an
> [Active/Active](/enterprise/install/automated/active-active)
> architecture.

Reviewer: ```suggestion
Storage independently of the Terraform Enterprise instance. Additionally, Active/Active stores Redis
data in an external Redis instance, allowing you to use multiple instances
of Terraform Enterprise in an
[Active/Active](/enterprise/install/automated/active-active)
architecture.
```

You are responsible for providing and maintaining the external PostgreSQL
server, S3-compatible object storage, and Redis instance.

> Configuration data is stored on a Docker volume on the instance. You can
> retrieve this information using `replicatedctl app-config export`.

Reviewer: ```suggestion
Active/Active stores configuration data on a Docker volume on the instance. Use `replicatedctl app-config export` to retrieve the configuration data.
```

### When to Use Active/Active

> Active/Active is recommended when:

Reviewer: ```suggestion
We recommend Active/Active for the following use cases:
```

> - Running Terraform Enterprise in environment that has an external PostgreSQL server, S3-compatible object storage, and Redis instance.

Reviewer: ```suggestion
- Your run environment that has an external PostgreSQL server, S3-compatible object storage, and Redis instance.
```

- You have experience managing a PostgreSQL server such as Amazon RDS for PostgreSQL, Google Cloud SQL for PostgreSQL, Azure Database for PostgreSQL, etc.
- You have experience managing an S3-compatible object storage location such as AWS S3, Azure Blob Storage, Google Cloud Storage, MinIO, etc.
- You have experience managing a Redis instance such as AWS ElastiCache for Redis, Azure Cache for Redis, Google Cloud Memorystore for Redis, etc.

> - Want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.
> - Want to scale beyond a single instance of Terraform Enterprise to increase reliability and performance.

Reviewer: ```suggestion
- You want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.
- You want to scale beyond a single instance of Terraform Enterprise to increase reliability and performance.
```

## Mounted Disk

The Mounted Disk operational mode stores the PostgreSQL Database and Object

> Storage data in a directory on the instance. It is expected that this directory
> is backed by persistent storage that can be mounted to the instance. This

Reviewer: ```suggestion
Storage data in a directory on the instance. It is expected that this directory
is backed by persistent storage that you can mount to the instance. This
```

allows you to backup and restore the PostgreSQL Database or Object Storage data
independently of the Terraform Enterprise instance.

You are responsible for providing and maintaining the persistent storage.

> Configuration data is stored on a Docker volume on the instance. You can
> retrieve this information using `replicatedctl app-config export`.

Reviewer: ```suggestion
Mounted Disk stores configuration data on a Docker volume on the instance. Use `replicatedctl app-config export` to retrieve the configuration data.
```

### When to Use Mounted Disk

Mounted Disk is recommended when:

> - Running Terraform Enterprise in an environment that does not have an external PostgreSQL server or S3-compatible object storage.
> - You have experience managing persistent storage such as an AWS EBS volume, an Azure Managed Disk, a Google Cloud Persistent Disk, an iSCSI location, etc.
> - Are familiar with using tools like `cp`, `scp`, and `rsync` to backup and restore data.

Reviewer: ```suggestion
- Your run environment does not have an external PostgreSQL server or S3-compatible object storage.
- You have experience managing persistent storage such as an AWS EBS volume, an Azure Managed Disk, a Google Cloud Persistent Disk, an iSCSI location, etc.
- You are familiar with using tools like `cp`, `scp`, and `rsync` to backup and restore data.
```

---
page_title: Operational Modes - Terraform Enterprise
description: >-
  Learn about the different operational modes of Terraform Enterprise.
---

# Terraform Enterprise Operational Modes

Before deploying Terraform Enterprise, you must choose an operational mode.
Each operational mode changes where Terraform Enterprise stores its data.

## Terraform Enterprise Data

Terraform Enterprise data utilizes the following types of data:

> - **PostgreSQL Database** - Stateful Terraform Enterprise application data. (e.g. workspace settings, organization settings, run information, user information)
> 
> - **Object Storage** - Artifacts generated by Terraform Enterprise. (e.g. state files, plan files, run logs, configuration versions)
> 
> - **Vault** - Encryption keys used to encrypt and decrypt objects within object storage.
> 
> - **Configuration** - Configuration settings. (e.g. the hostname, object storage credentials, database credentials, concurrency settings, etc.)

Reviewer: ```suggestion
- **PostgreSQL Database:** - Stateful Terraform Enterprise application data. This includes workspace settings, organization settings, run information, and user information.

- **Object Storage:** Artifacts that Terraform Enterprise produces during operations. This includes state files, plan files, run logs, configuration versions.

- **Vault:** Encryption keys that encrypt and decrypt objects within object storage.

- **Configuration:** Configuration settings. This includes the hostname, object storage credentials, database credentials, concurrency settings, etc.
```

Reviewer: should we also link to the PostgreSQL requirements page? https://www.terraform.io/enterprise/requirements/data-storage/postgres-requirements

Author: I don't think so. We're just listing the types of data here.

## Operational Modes

> The following table lists each operational mode and where it stores each type of data.

Reviewer: ```suggestion
The following table summarizes where each Terraform Enterprise operational mode stores each type of application data. 
```

| Operational Mode                        | PostgreSQL Database                                                              | Object Storage                                                                  | Vault                                                                             | Configuration                              |
| --------------------------------------- | -------------------------------------------------------------------------------- | ------------------------------------------------------------------------------- | --------------------------------------------------------------------------------- | ------------------------------------------ |

> | [Mounted Disk](#mounted-disk)           | Stored in a directory on the instance backed by user-managed persistent storage. | Stored in a directory on the instance backed by user-managed persisent storage. | Stored in the PostgreSQL Database unless using [External Vault](#external-vault). | Stored in a Docker volume on the instance. |

Reviewer: is there a way to take out the "stored in" in the front of each of these? It'd be easier to parse if we got more to the actual point quickly. So like "Directory on the Instance"  "PostgreSQL database, unless using External Vault." etc. Thoughts?

Author: Done!

## External Services

The External Services operational mode stores the PostgreSQL Database and Object Storage data on an external PostgreSQL server and S3-compatible object storage. This allows you to backup, restore, and scale the PostgreSQL Database or Object Storage independently of the Terraform Enterprise instance.

You are responsible for providing and maintaining the external PostgreSQL server and S3-compatible object storage.

Configuration data is stored on a Docker volume on the instance. You can retrieve this information using `replicatedctl app-config export`.

### When to Use External Services

External Services is recommended when:

- Running Terraform Enterprise in environment that has an external PostgreSQL server or S3-compatible object storage.
- You have experience managing a PostgreSQL server such as Amazon RDS for PostgreSQL, Google Cloud SQL for PostgreSQL, Azure Database for PostgreSQL, etc.
- You have experience managing an S3-compatible object storage location such as AWS S3, Azure Blob Storage, Google Cloud Storage, MinIO, etc.
- Want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.

> ## External Vault

Reviewer: This is not currently on the pre-install checklist (there we have active/active, mounted disk, and external services). Should it be? I'm confused about the actual complete list of all the modes (is active/active one? Is External vault one?). Can we make sure everything is consistent across these pages?  Please forgive me if this is my fault when I did the rework of the sidebar!

rAnd then, I'm wondering if we can cut some of this because we have this nice new page now :) : https://www.terraform.io/enterprise/requirements/data-storage/vault

Author: External Vault is not an operational mode in itself. It can be used across the different operational modes. I removed the information and updated the new page content.

---
page_title: Operational Modes - Terraform Enterprise
description: >-
  TODO
---

# Terraform Enterprise Operational Modes

Before deploying Terraform Enterprise, you must choose an operational mode.
Each operational mode changes where Terraform Enterprise stores its data.

> ## Terraform Enterprise Data

Reviewer 2: Sort of both. They are the primary groups/types of data that TFE uses in order to function. They are different from one another in type but all are required for TFE to work properly
Their location is entirely dependent on the operational mode, so I agree that its important for the customer to have a firm idea of WHAT they are and WHERE they will be depending on the operational mode they choose.

> ## Terraform Enterprise Data

Author: What @duplo83 said. Terraform Enterprise has application data and configuration. The application data is spread across PostgreSQL, Object Storage, and Vault. The configuration is the settings file that users use to configure Terraform Enterprise. An operational mode is essentially just a specific combination of where application data is stored (i.e. mounted disk means application data lives on the host itself in a directory).

Terraform Enterprise data utilizes the following types of data:

- **PostgreSQL Database** - Stateful Terraform Enterprise application data. (e.g. workspace settings, organization settings, run information, user information)

- **Object Storage** - Artifacts generated by Terraform Enterprise. (e.g. state files, plan files, run logs, configuration versions)

- **Vault** - Encryption keys used to encrypt and decrypt objects within object storage.

- **Configuration** - Configuration settings. (e.g. the hostname, object storage credentials, database credentials, concurrency settings, etc.)

> ## Operational Modes

Reviewer 2: I like it. The data column title could possibly contain the definitions of the data types?

> ## Operational Modes

Author: Just to be clear @laurapacilio, you're suggesting to combine the `Terraform Enterprise Data` and `Operational Modes` sections?

> ## Operational Modes

Author: Going to resolve this conversation in favor of the other one.

> ## Operational Modes

Reviewer: you know what? I take this back. Reading it again I don't know what I meant by this anymore

| [Mounted Disk](#mounted-disk)           | Stored in a directory on the instance backed by user-managed persistent storage. | Stored in a directory on the instance backed by user-managed persisent storage. | Stored in the PostgreSQL Database unless using [External Vault](#external-vault). | Stored in a Docker volume on the instance. |
| [External Services](#external-services) | Stored in an external user-managed PostgreSQL database.                          | Stored in an external user-managed object storage location.                     | Stored in the PostgreSQL Database unless using [External Vault](#external-vault). | Stored in a Docker volume on the instance. |

## Demo

!> **Warning:** Do not use the Demo operational mode in a production environment. Terraform Enterprise data is lost when the instance terminates.

The Demo operational mode stores all data (PostgreSQL Database, Object Storage, Configuration) in Docker volumes on the instance. This allows you to quickly deploy Terraform Enterprise without any additional infrastructure. However, when the instance terminates, all data is lost.

The Demo operational mode is best when you:

> - Are okay with all the data being lost when the instance terminates.

Reviewer 2: ```suggestion
- Understand that all data (PostgreSQL Database, Object Storage, Configuration) will be lost when the instance terminates.
```
It might be best to fully clarify, again, what is at stake if they choose demo mode.

Reviewer: +1. I actually even wonder if we should change this slightly and then put this callout in a callout for emphasis. So like this bullet would say "Do not need to keep the data permanently" and then provide an example of when that would be. And then have a separate callout somewhere like:

~> **Note:** All data will be lost when the instance terminates." 

Maybe too much, but I want to make sure people don't use this unless it's really appropriate.

Author: The demo operational mode is gone now so I get to do the best choice and remove this altogether!

- Are deploying Terraform Enterprise to a non-production, proof of concept environment.
- Do not want to manage additional infrastructure besides an instance.

## Mounted Disk

The Mounted Disk operational mode stores the PostgreSQL Database and Object Storage data in a directory on the instance. It is expected that this directory is backed by persistent storage that can be mounted to the instance. This allows you to backup and restore the PostgreSQL Database or Object Storage data independently of the Terraform Enterprise instance.

You are responsible for providing and maintaining the persistent storage.

Configuration data is stored on a Docker volume on the instance. You can retrieve this information using `replicatedctl app-config export`.

> The Mounted Disk operational mode is best when you:

Author: Updated!

Configuration data is stored on a Docker volume on the instance. You can retrieve this information using `replicatedctl app-config export`.

The External Services operational mode is best when you:

- Want PostgreSQL Database and Object Storage data to persist even if the instance is terminated.
- Have experience managing a PostgreSQL server such as Amazon RDS for PostgreSQL, Google Cloud SQL for PostgreSQL, Azure Database for PostgreSQL, etc.
- Have experience managing an S3-compatible object storage location such as AWS S3, Azure Blob Storage, Google Cloud Storage, MinIO, etc.
- Want to use native backup and restore features provided by the external PostgreSQL server and S3-compatible object storage.

## External Vault

Terraform Enterprise automatically creates an internally-managed Vault server that stores its data in the PostgreSQL Database. We strongly recommend that organizations use this internally-managed Vault server. However, some organizations have specific requirements around data encryption and auditing. Those organizations can configure Terraform Enterprise to use an external Vault server rather than the internally-managed Vault server.

~> **Important:** TODO

> !> **Warning:** TODO

Author: Updated!
